### PR TITLE
Add MBPP evaluator

### PR DESCRIPTION
**Pull request by natke**
## Summary
Adds MBPP evaluator support for the `ortgenai` backend, enabling generative LM eval tasks (e.g., MBPP, HumanEval) via `lm-eval`. Key changes include:

- Add `generate_until` support to `LMEvalORTGenAIEvaluator` for token-by-token generation with stop-sequence handling
- Wire `confirm_run_unsafe_code` opt-in flag from CLI through to `lm_eval.simple_evaluate` for MBPP safety compliance
- Improve stop-sequence trimming to cut at the earliest match across all provided stop strings
- Improve `generate_until` with full-sequence decode for correct tokenizer whitespace/punctuation handling
- Harden `gen_kwargs` parsing: max tokens, temperature, and `do_sample` coercion for string/int inputs
- Add evaluator tests for unsafe-code flag propagation, multi-stop ordering, and token parsing edge cases

## Example Olive config

The evaluator can be used standalone against a pre-existing ONNX model — no optimization passes are required. The `passes` field defaults to empty, and `evaluate_input_model` defaults to `true`, so Olive evaluates the input model directly:

```json
{
    "input_model": {
        "type": "OnnxModel",
        "model_path": "models/Qwen2.5-Coder-0.5B-Instruct-cpu-int4"
    },
    "systems": {
        "local_system": {
            "type": "LocalSystem",
            "accelerators": [{ "device": "cpu", "execution_providers": ["CPUExecutionProvider"] }]
        }
    },
    "evaluators": {
        "lm_evaluator": {
            "type": "LMEvaluator",
            "tasks": ["mbpp"],
            "model_class": "ortgenai",
            "batch_size": 1,
            "max_length": 2048,
            "limit": 10,
            "confirm_run_unsafe_code": true
        }
    },
    "target": "local_system",
    "evaluator": "lm_evaluator"
}
```

Run with:
```bash
olive run -c config.json
```

Or via the CLI directly (no config file needed):
```bash
olive benchmark -m models/Qwen2.5-Coder-0.5B-Instruct-cpu-int4 \
    --tasks mbpp \
    --backend ortgenai \
    --confirm_run_unsafe_code
```

> **Note:** MBPP executes generated Python code to check correctness. The `confirm_run_unsafe_code` flag (or `--confirm_run_unsafe_code` on the CLI) is required to acknowledge this.

## Validation
- pytest -q test/evaluator/test_olive_evaluator.py -k "generate_until or confirm_run_unsafe_code or lmeval"
- pytest -q test/cli/test_cli.py -k "benchmark_command_onnxmodel_with_ortgenai_backend or benchmark_command_hfmodel"

### CODE REVIEW

File: olive/evaluator/lmeval_ort.py [hunk 1]
```diff
@@ -573,5 +583,93 @@ def model_call(self, input_ids: torch.Tensor, cont_len: int = 0) -> torch.Tensor
...
+            else:
+                search_options["temperature"] = 0.0
+            params.set_search_options(**search_options)
+
+            # Run generation token by token to check for stop sequences
+            generator = og.Generator(self.model, params)
+            generator.append_tokens([prompt_ids])
+
+            generated_ids = []
+            generated_text = ""
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/lmeval_ort.py]: Fixed in commit a3e3930f — removed the dead `generated_ids` list and replaced `generated_text +=` string concatenation with `generated_chunks` list accumulation and `"".join()`.

File: olive/evaluator/lmeval_ort.py [hunk 2]
```diff
@@ -573,5 +583,93 @@ def model_call(self, input_ids: torch.Tensor, cont_len: int = 0) -> torch.Tensor
...
+                generator.generate_next_token()
+                new_token = generator.get_sequence(0)[-1]
+
+                # Check for EOS token(s)
+                if new_token in self._eos_token_ids:
+                    break
+
+                generated_ids.append(new_token)
+                generated_text += self.tokenizer.decode([new_token])
+
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/lmeval_ort.py]: Fixed in commit a3e3930f — removed the dead `generated_ids` list and replaced `generated_text +=` string concatenation with `generated_chunks` list accumulation and `"".join()`.

File: test/evaluator/test_olive_evaluator.py [hunk 3]
```diff
@@ -510,3 +510,310 @@
...
+        return np.array(ids)
+
+    @patch("onnxruntime_genai.Generator")
+    @patch("onnxruntime_genai.GeneratorParams")
+    def test_generate_until_stops_on_eos(self, mock_params_cls, mock_gen_cls):
+        """Test that generation stops when EOS token is produced."""
+        from olive.evaluator.lmeval_ort import LMEvalORTGenAIEvaluator
+
+        evaluator = MagicMock(spec=LMEvalORTGenAIEvaluator)
+        evaluator._eos_token_ids = {2}
```
[the remark this answers is not in the document]
Comment (natke) [reply test/evaluator/test_olive_evaluator.py]: Fixed in commit a3e3930f — renamed `_eos_token_ids` to the public attribute `eos_token_ids` in the class definition and all usages, removing the protected-access violation.

File: test/evaluator/test_olive_evaluator.py [hunk 4]
```diff
@@ -510,3 +510,310 @@
...
+        assert results[0] == "hello"
+
+    @patch("onnxruntime_genai.Generator")
+    @patch("onnxruntime_genai.GeneratorParams")
+    def test_generate_until_stops_on_stop_sequence(self, mock_params_cls, mock_gen_cls):
+        """Test that generation stops and trims at stop sequence."""
+        from olive.evaluator.lmeval_ort import LMEvalORTGenAIEvaluator
+
+        evaluator = MagicMock(spec=LMEvalORTGenAIEvaluator)
+        evaluator._eos_token_ids = {2}
```
[the remark this answers is not in the document]
Comment (natke) [reply test/evaluator/test_olive_evaluator.py]: Fixed in commit a3e3930f — renamed `_eos_token_ids` to the public attribute `eos_token_ids` in the class definition and all usages, removing the protected-access violation.

File: test/evaluator/test_olive_evaluator.py [hunk 5]
```diff
@@ -510,3 +510,310 @@
...
+        assert results[0] == "hello"  # trimmed at \n
+
+    @patch("onnxruntime_genai.Generator")
+    @patch("onnxruntime_genai.GeneratorParams")
+    def test_generate_until_respects_max_length(self, mock_params_cls, mock_gen_cls):
+        """Test that total_max_length = min(prompt_len + max_gen_toks, max_length)."""
+        from olive.evaluator.lmeval_ort import LMEvalORTGenAIEvaluator
+
+        evaluator = MagicMock(spec=LMEvalORTGenAIEvaluator)
+        evaluator._eos_token_ids = {2}
```
[the remark this answers is not in the document]
Comment (natke) [reply test/evaluator/test_olive_evaluator.py]: Fixed in commit a3e3930f — renamed `_eos_token_ids` to the public attribute `eos_token_ids` in the class definition and all usages, removing the protected-access violation.

File: test/evaluator/test_olive_evaluator.py [hunk 6]
```diff
@@ -510,3 +510,310 @@
...
+        assert call_kwargs["max_length"] == 50
+
+    @patch("onnxruntime_genai.Generator")
+    @patch("onnxruntime_genai.GeneratorParams")
+    def test_generate_until_handles_multiple_eos_tokens(self, mock_params_cls, mock_gen_cls):
+        """Test that any token in _eos_token_ids triggers stop."""
+        from olive.evaluator.lmeval_ort import LMEvalORTGenAIEvaluator
+
+        evaluator = MagicMock(spec=LMEvalORTGenAIEvaluator)
+        evaluator._eos_token_ids = {2, 151645, 151643}  # Multiple EOS like Qwen
```
[the remark this answers is not in the document]
Comment (natke) [reply test/evaluator/test_olive_evaluator.py]: Fixed in commit a3e3930f — renamed `_eos_token_ids` to the public attribute `eos_token_ids` in the class definition and all usages, removing the protected-access violation.

File: test/evaluator/test_olive_evaluator.py [hunk 7]
```diff
@@ -510,3 +510,310 @@
...
+
+        assert len(results) == 1
+        assert results[0] == "result"
+
+    def test_generate_until_until_string_converted_to_list(self):
+        """Test that a string 'until' value is converted to a list."""
+        from olive.evaluator.lmeval_ort import LMEvalORTGenAIEvaluator
+
+        evaluator = MagicMock(spec=LMEvalORTGenAIEvaluator)
+        evaluator._eos_token_ids = {2}
```
[the remark this answers is not in the document]
Comment (natke) [reply test/evaluator/test_olive_evaluator.py]: Fixed in commit a3e3930f — renamed `_eos_token_ids` to the public attribute `eos_token_ids` in the class definition and all usages, removing the protected-access violation.

File: test/evaluator/test_olive_evaluator.py [hunk 8]
```diff
@@ -510,3 +510,310 @@
...
+            assert "\n" not in results[0]
+
+    @patch("onnxruntime_genai.Generator")
+    @patch("onnxruntime_genai.GeneratorParams")
+    def test_generate_until_uses_earliest_stop_match(self, mock_params_cls, mock_gen_cls):
+        """Test that stop trimming uses earliest occurrence across all stop sequences."""
+        from olive.evaluator.lmeval_ort import LMEvalORTGenAIEvaluator
+
+        evaluator = MagicMock(spec=LMEvalORTGenAIEvaluator)
+        evaluator._eos_token_ids = {2}
```
[the remark this answers is not in the document]
Comment (natke) [reply test/evaluator/test_olive_evaluator.py]: Fixed in commit a3e3930f — renamed `_eos_token_ids` to the public attribute `eos_token_ids` in the class definition and all usages, removing the protected-access violation.

File: test/evaluator/test_olive_evaluator.py [hunk 9]
```diff
@@ -510,3 +510,310 @@
...
+    @patch("onnxruntime_genai.Generator")
+    @patch("onnxruntime_genai.GeneratorParams")
+    def test_generate_until_parses_max_tokens_robustly(
+        self, mock_params_cls, mock_gen_cls, gen_kwargs, expected_max_length
+    ):
+        """Test robust parsing and clamping of max token kwargs."""
+        from olive.evaluator.lmeval_ort import LMEvalORTGenAIEvaluator
+
+        evaluator = MagicMock(spec=LMEvalORTGenAIEvaluator)
+        evaluator._eos_token_ids = {2}
```
[the remark this answers is not in the document]
Comment (natke) [reply test/evaluator/test_olive_evaluator.py]: Fixed in commit a3e3930f — renamed `_eos_token_ids` to the public attribute `eos_token_ids` in the class definition and all usages, removing the protected-access violation.

File: test/evaluator/test_olive_evaluator.py [hunk 10]
```diff
@@ -510,3 +510,310 @@
...
+        assert call_kwargs["max_length"] == expected_max_length
+
+    @patch("onnxruntime_genai.Generator")
+    @patch("onnxruntime_genai.GeneratorParams")
+    def test_generate_until_decodes_incrementally(self, mock_params_cls, mock_gen_cls):
+        """Test generation decodes only new tokens while preserving output."""
+        from olive.evaluator.lmeval_ort import LMEvalORTGenAIEvaluator
+
+        evaluator = MagicMock(spec=LMEvalORTGenAIEvaluator)
+        evaluator._eos_token_ids = {2}
```
[the remark this answers is not in the document]
Comment (natke) [reply test/evaluator/test_olive_evaluator.py]: Fixed in commit a3e3930f — renamed `_eos_token_ids` to the public attribute `eos_token_ids` in the class definition and all usages, removing the protected-access violation.

File: olive/evaluator/lmeval_ort.py [hunk 11]
```diff
@@ -573,5 +583,93 @@ def model_call(self, input_ids: torch.Tensor, cont_len: int = 0) -> torch.Tensor
...
+            except (TypeError, ValueError):
+                max_gen_toks = 256
+            max_gen_toks = max(max_gen_toks, 0)
+            temperature = gen_kwargs.get("temperature", 0.0)
+            do_sample = gen_kwargs.get("do_sample", temperature > 0)
+
+            # Tokenize the prompt
+            prompt_ids = self.tokenizer.encode(context).tolist()
+            prompt_len = len(prompt_ids)
+
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/lmeval_ort.py:634]: Fixed in commit 33f93343 — added an early-return path before the generator is created: when `prompt_len >= self.max_length` or `max_gen_toks == 0`, the function appends an empty completion and continues to the next request, so the generator is never given an invalid `max_length`.

File: olive/evaluator/lmeval_ort.py [hunk 12]
```diff
@@ -509,7 +512,14 @@ def __init__(
...
                 self.max_length = genai_config["search"]["max_length"]
-            self._eot_token_id = genai_config["model"]["eos_token_id"]
+            eos = genai_config["model"]["eos_token_id"]
+            # eos_token_id can be a single int or a list of ints
+            if isinstance(eos, list):
+                self._eot_token_id = eos[0]
+                self.eos_token_ids = set(eos)
+            else:
+                self._eot_token_id = eos
+                self.eos_token_ids = {eos}
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/lmeval_ort.py]: Fixed in commit 33f93343 — added a guard before accessing `eos[0]`: if the list is empty a `ValueError` is raised with a clear message rather than crashing with an `IndexError` at runtime.

File: olive/evaluator/lmeval_ort.py [hunk 13]
```diff
@@ -573,5 +583,93 @@ def model_call(self, input_ids: torch.Tensor, cont_len: int = 0) -> torch.Tensor
...
+                # Check stop sequences against generated text
+                earliest_stop_idx = None
+                for stop_seq in until:
+                    stop_idx = generated_text.find(stop_seq)
+                    if stop_idx != -1 and (earliest_stop_idx is None or stop_idx < earliest_stop_idx):
+                        earliest_stop_idx = stop_idx
+
+                if earliest_stop_idx is not None:
+                    generated_text = generated_text[:earliest_stop_idx]
+                    break
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/lmeval_ort.py]: Fixed in commit 33f93343 — replaced the full-join-per-token approach with a bounded tail-buffer of width `max(len(s) for s in until)`. Stop sequences are checked against this tail on each token; the full output list is joined only once after the loop exits (or on a stop-sequence match), keeping the hot path O(1) per token.

File: olive/evaluator/lmeval_ort.py [hunk 14]
```diff
@@ -573,5 +585,108 @@ def model_call(self, input_ids: torch.Tensor, cont_len: int = 0) -> torch.Tensor
...
+                if new_token in self.eos_token_ids:
+                    break
+
+                chunk = self.tokenizer.decode([new_token])
+                generated_chunks.append(chunk)
+
+                # Check stop sequences against a tail window to avoid O(n²) full join
+                if until:
+                    tail = "".join(generated_chunks[-(max_stop_len + 1) :]) if max_stop_len else ""
+                    tail_offset = len("".join(generated_chunks)) - len(tail)
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/lmeval_ort.py]: Fixed in commit 957d755c — replaced `len("".join(generated_chunks))` with a `generated_len` running counter that increments by `len(chunk)` each token, so `tail_offset` is computed in O(1) with no string joining.

File: olive/evaluator/lmeval_ort.py [hunk 15]
```diff
@@ -573,5 +585,108 @@ def model_call(self, input_ids: torch.Tensor, cont_len: int = 0) -> torch.Tensor
...
+            params.set_search_options(**search_options)
+
+            # Run generation token by token to check for stop sequences
+            generator = og.Generator(self.model, params)
+            generator.append_tokens([prompt_ids])
+
+            generated_chunks = []
+            stop_idx = None
+            # Tail buffer wide enough to detect any stop sequence across chunk boundaries
+            max_stop_len = max((len(s) for s in until), default=0)
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/lmeval_ort.py]: Fixed in commit 957d755c — replaced the chunk-count slice (`generated_chunks[-N:]`) with a character-based rolling `tail` string maintained as `(tail + chunk)[-(max_stop_len + len(chunk)):]`. This ensures stop sequences are never missed regardless of how many tokens they span or how short each decoded chunk is.

File: olive/evaluator/lmeval_ort.py [hunk 16]
```diff
@@ -573,5 +585,108 @@ def model_call(self, input_ids: torch.Tensor, cont_len: int = 0) -> torch.Tensor
...
+                # Check for EOS token(s)
+                if new_token in self.eos_token_ids:
+                    break
+
+                chunk = self.tokenizer.decode([new_token])
+                generated_chunks.append(chunk)
+
+                # Check stop sequences against a tail window to avoid O(n²) full join
+                if until:
+                    tail = "".join(generated_chunks[-(max_stop_len + 1) :]) if max_stop_len else ""
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/lmeval_ort.py]: Fixed in commit 957d755c — same fix as above: the tail is now a rolling character string, not a chunk-index slice.

File: test/evaluator/test_olive_evaluator.py [hunk 17]
```diff
@@ -510,3 +510,328 @@ def test_lm_evaluator_dispatches_to_requested_backend(
...
+        assert results[0] == "hello"
+
+    @pytest.mark.parametrize(
+        ("gen_kwargs", "expected_max_length"),
+        [
+            (None, 261),  # default 256 when gen_kwargs is not a dict
+            ({"max_gen_toks": "7"}, 12),  # parse numeric string
+            ({"max_new_tokens": "bad"}, 261),  # invalid value falls back to default
+        ],
+    )
```
[the remark this answers is not in the document]
Comment (natke) [reply test/evaluator/test_olive_evaluator.py:837]: Fixed in commit 957d755c — `temperature` is now coerced with `float(...)` inside a `try/except (TypeError, ValueError)` with fallback to `0.0`, so string values and `None` are handled safely. Added a 4-case parametrized test covering `"0.7"`, `None`, `0.0`, and `0.5`.

File: olive/evaluator/lmeval_ort.py [hunk 18]
```diff
@@ -573,5 +585,116 @@ def model_call(self, input_ids: torch.Tensor, cont_len: int = 0) -> torch.Tensor
...
+            search_options = {
+                "max_length": total_max_length,
+                "past_present_share_buffer": False,
+                "batch_size": 1,
+            }
+            if do_sample:
+                search_options["temperature"] = temperature
+            else:
+                search_options["temperature"] = 0.0
+            params.set_search_options(**search_options)
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/lmeval_ort.py:656]: Fixed in commit b6e89d1d. Removed `batch_size` from the `set_search_options` call — it is not a valid keyword argument for ORT GenAI's `GeneratorParams`.

File: olive/evaluator/olive_evaluator.py [hunk 19]
```diff
@@ -1108,6 +1109,7 @@ def evaluate(
                 batch_size=self.batch_size,
                 device=device,
                 limit=self.limit,
+                confirm_run_unsafe_code=self.confirm_run_unsafe_code,
             )
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/olive_evaluator.py]: Fixed in commit b6e89d1d. Switched to a `try/except TypeError` pattern: the kwargs dict is built with `confirm_run_unsafe_code` included, and if `simple_evaluate` raises `TypeError` mentioning that parameter, it is popped and the call is retried — maintaining compatibility with older `lm-eval` versions.

File: olive/evaluator/lmeval_ort.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/lmeval_ort.py:656]: Updated in 0a3a489d: switched to `inspect.signature(simple_evaluate).parameters` to proactively check whether the installed lm-eval version supports `confirm_run_unsafe_code` before passing it. Added a test that simulates an older lm-eval without that parameter to confirm it is excluded.

File: olive/evaluator/lmeval_ort.py [hunk 20]
```diff
@@ -573,5 +585,117 @@ def model_call(self, input_ids: torch.Tensor, cont_len: int = 0) -> torch.Tensor
...
+                generator.generate_next_token()
+                new_token = generator.get_sequence(0)[-1]
+
+                # Check for EOS token(s)
+                if new_token in self.eos_token_ids:
+                    break
+
+                chunk = self.tokenizer.decode([new_token])
+                generated_chunks.append(chunk)
+                generated_len += len(chunk)
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/lmeval_ort.py]: Fixed in 0a3a489d: token IDs are now accumulated in `generated_token_ids`. In the loop, per-token decode is still done (but only when stop sequences are present) for the rolling tail buffer to enable early exit. After the loop, the full sequence is decoded once with `self.tokenizer.decode(generated_token_ids)` so that tokenizer whitespace/punctuation normalisation is applied correctly. Stop sequences are then found by searching the final decoded string.

File: olive/evaluator/lmeval_ort.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/lmeval_ort.py:656]: Fixed in 0a3a489d. Replaced the `generated_chunks` list and per-token `''.join()` with accumulated `generated_token_ids`. Per-token decode is still done inside the loop when stop sequences are present (for the rolling tail buffer), but the final `generated_text` is now produced by `self.tokenizer.decode(generated_token_ids)` — a single full-sequence decode — so tokenizer whitespace/punctuation normalisation applies correctly.

File: olive/evaluator/olive_evaluator.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/olive_evaluator.py]: Fixed in 0a3a489d: switched from `try/except TypeError` to `inspect.signature(simple_evaluate).parameters` — proactively checks the installed lm-eval's function signature and only includes `confirm_run_unsafe_code` when the parameter is present. No message-string matching needed, no unrelated TypeErrors can be masked.

File: olive/evaluator/lmeval_ort.py [hunk 21]
```diff
@@ -573,5 +585,117 @@ def model_call(self, input_ids: torch.Tensor, cont_len: int = 0) -> torch.Tensor
...
+            try:
+                max_gen_toks = int(max_gen_toks) if max_gen_toks is not None else 256
+            except (TypeError, ValueError):
+                max_gen_toks = 256
+            max_gen_toks = max(max_gen_toks, 0)
+            try:
+                temperature = float(gen_kwargs.get("temperature", 0.0) or 0.0)
+            except (TypeError, ValueError):
+                temperature = 0.0
+            do_sample = gen_kwargs.get("do_sample", temperature > 0)
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/lmeval_ort.py]: Fixed in 0a3a489d: `do_sample` is now coerced properly. Booleans pass through; strings are compared against `('false', '0', 'no', '')` for falsiness; other types use `bool()`. A parametrized test covers string `'false'`/`'0'`, int `0`/`1`, and plain booleans.

File: olive/evaluator/lmeval_ort.py [hunk 22]
```diff
@@ -573,5 +585,117 @@ def model_call(self, input_ids: torch.Tensor, cont_len: int = 0) -> torch.Tensor
...
+                        idx = tail.find(stop_seq)
+                        if idx != -1:
+                            abs_idx = tail_offset + idx
+                            if earliest is None or abs_idx < earliest:
+                                earliest = abs_idx
+                    if earliest is not None:
+                        stop_idx = earliest
+                        break
+
+            generated_text = "".join(generated_chunks) if stop_idx is None else "".join(generated_chunks)[:stop_idx]
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/lmeval_ort.py]: Fixed in 0a3a489d: the whole `generated_chunks` approach was replaced by accumulating token IDs and calling `self.tokenizer.decode(generated_token_ids)` once after the loop — no join at all.

[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/lmeval_ort.py]: Fixed in 0a3a489d (same fix as the other thread). Token IDs are accumulated and the full sequence is decoded once after the loop with `self.tokenizer.decode(generated_token_ids)`.

File: olive/evaluator/lmeval_ort.py [hunk 23]
```diff
@@ -573,5 +585,131 @@ def model_call(self, input_ids: torch.Tensor, cont_len: int = 0) -> torch.Tensor
...
+
+            # Extract stop sequences
+            until = gen_kwargs.get("until", [])
+            if isinstance(until, str):
+                until = [until]
+            elif until is None:
+                until = []
+            elif not isinstance(until, list):
+                until = [until]
+            until = [stop_seq for stop_seq in until if isinstance(stop_seq, str) and stop_seq]
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/lmeval_ort.py]: Fixed in 0b62f2d0: changed `elif not isinstance(until, list): until = [until]` to `until = list(until)` inside a `try/except TypeError` so tuples, sets, and other iterables are unpacked rather than wrapped. Added a regression test (`test_generate_until_handles_tuple_until`) to confirm `until=('\\n',)` still trims output correctly.

File: test/evaluator/test_olive_evaluator.py [hunk 24]
```diff
@@ -510,3 +516,504 @@ def test_lm_evaluator_dispatches_to_requested_backend(
...
+        evaluator.evaluate(model, metrics=[], device=Device.CPU, execution_providers=["CPUExecutionProvider"])
+
+        call_kwargs = simple_evaluate_mock.call_args[1]
+        assert "confirm_run_unsafe_code" not in call_kwargs
+
+
+@pytest.mark.skipif(
+    importlib.util.find_spec("lm_eval") is None or importlib.util.find_spec("onnxruntime_genai") is None,
+    reason="lm_eval or onnxruntime_genai not installed",
+)
```
[the remark this answers is not in the document]
Comment (natke) [reply test/evaluator/test_olive_evaluator.py:641]: Fixed in 0b62f2d0: added `test/evaluator/conftest.py` which injects a minimal `onnxruntime_genai` stub into `sys.modules` when the real package is absent. The `TestLMEvalORTGenAIGenerateUntil` skip condition now only guards on `lm_eval`, so the unit tests run in any standard CI environment.

File: olive/evaluator/olive_evaluator.py [hunk 25]
```diff
@@ -1100,15 +1102,19 @@ def evaluate(
...
+                "model": lmmodel,
+                "tasks": self.tasks,
+                "task_manager": TaskManager(),
+                "log_samples": False,
+                "batch_size": self.batch_size,
+                "device": device,
+                "limit": self.limit,
+            }
+            # Only pass confirm_run_unsafe_code when the installed lm-eval version supports it.
+            if "confirm_run_unsafe_code" in inspect.signature(simple_evaluate).parameters:
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/evaluator/olive_evaluator.py]: Fixed in 0b62f2d0: wrapped the signature check in `try/except (TypeError, ValueError)` with `supports_confirm_run_unsafe_code = False` as the fallback, exactly as suggested.

File: olive/cli/benchmark.py [hunk 26]
```diff
@@ -117,6 +124,10 @@ def _get_run_config(self, tempdir: str) -> dict:
                 ("evaluators", "evaluator", "model_class"),
                 None if self.args.backend == "auto" else self.args.backend,
             ),
+            (
+                ("evaluators", "evaluator", "confirm_run_unsafe_code"),
+                self.args.confirm_run_unsafe_code or None,
```
[the remark this answers is not in the document]
Comment (natke) [reply olive/cli/benchmark.py]: Fixed in 0b62f2d0: changed `self.args.confirm_run_unsafe_code or None` to `self.args.confirm_run_unsafe_code` directly. Since `action='store_true'` always gives a bool, `False is not None` → the value is now always written explicitly as `true` or `false`.

**Comment by jambayk** [comment]
@copilot resolve the merge conflicts in this pull request

**Comment by jambayk** [comment]
@copilot resolve the merge conflicts in this pull request

**Comment by jambayk** [comment]
@copilot resolve the merge conflicts in this pull request

**Comment by jambayk** [comment]
closing since this has become stale
